stream.go

```diff
@@ -0,0 +1,120 @@
+package dht
+
+import (
+	"context"
+	"sync"
+
+	pbio "github.com/gogo/protobuf/io"
+	pb "github.com/libp2p/go-libp2p-kad-dht/pb"
+	inet "github.com/libp2p/go-libp2p-net"
+	peer "github.com/libp2p/go-libp2p-peer"
+	"golang.org/x/xerrors"
+)
+
+func (dht *IpfsDHT) newStream(ctx context.Context, p peer.ID) (*stream, error) {
+	s, err := dht.host.NewStream(ctx, p, dht.protocols...)
+	if err != nil {
+		return nil, xerrors.Errorf("opening stream: %w", err)
+	}
+	ps := &stream{
+		stream: s,
+		w:      newBufferedDelimitedWriter(s),
+		r:      pbio.NewDelimitedReader(s, inet.MessageSizeMax),
+		m:      make(chan chan *pb.Message, 1),
+	}
+	go func() {
+		ps.reader()
+		dht.streamPool.delete(ps, p)
+		ps.reset()
+	}()
+	return ps, nil
+}
+
+type stream struct {
+	stream interface {
+		Reset() error
+	}
+	w bufferedWriteCloser
+	r pbio.ReadCloser
+
+	// Synchronizes m and readerErr.
+	mu sync.Mutex
+	// Receives channels to send responses on.
+	m         chan chan *pb.Message
+	readerErr error
+}
+
+func (me *stream) reset() {
+	me.stream.Reset()
+}
+
+func (me *stream) send(m *pb.Message) (err error) {
+	if err := me.w.WriteMsg(m); err != nil {
+		return xerrors.Errorf("writing message: %w", err)
+	}
+	if err := me.w.Flush(); err != nil {
+		return xerrors.Errorf("flushing: %w", err)
+	}
+	return nil
+}
+
+func (me *stream) request(ctx context.Context, req *pb.Message) (<-chan *pb.Message, error) {
+	replyChan := make(chan *pb.Message, 1)
+	me.mu.Lock()
+	if err := me.errLocked(); err != nil {
+		me.mu.Unlock()
+		return nil, err
+	}
+	select {
+	case me.m <- replyChan:
+	default:
+		me.mu.Unlock()
+		panic("message pipeline full")
```

Comment:
I don't think we should kill the entire system here if a stream is double booked. An error should be enough.

Reply:
Definitely not! That's not a good reason to panic.

Reply:
It's a logical error. If this ever happened, assumptions in the code base are incorrect. For example, the stream has been picked up by 2 users at the same time, or we sent 2 messages without getting a reply for one.

Reply:
Can you at least add some comments for the reasons of the panic?

Reply:
-1 to panicking. Killing the entire system is a horrible way of reporting that an assumption didn’t stand.

Reply:
I'm actually with @anacrolix here. If something goes wrong here, the system is in an unknown state (possible due to an unrelated memory corruption); panicing is the standard way to handle kind of situation (granted, that's usually followed by a recover at a request boundary...). Error logs are useful for cases we don't _think_ should happen but this is a case that really can't happen.

Reply:
That's fine, but can we please add a comment there for the reason to panic and why it is justified?

Reply:
Added